### PR TITLE
fix: fix an ordering issue that blocks pnr

### PR DESCRIPTION
**Pull request by tianrui-wei**
This commit fixes an ordering issue such that under these scenarios, PNR will be broken.

1. an older load misses
2. a younger load gets issued after the cacheline comes back but before the older load gets replayed
3. immediately after the younger load hits, the line gets evicted due to snooping.

This breaks ld-ld ordering for the same address, and although the old line conceives no memory ordering violation could be made, it still needs to be replayed, thus blocking PnR.

**Related issue**: 

**Type of change**: bug fix

**Impact**: rtl refactoring

**Development Phase**: implementation

**Release Notes**

### CODE REVIEW

File: src/main/scala/v4/lsu/lsu.scala [hunk 1]
```diff
@@ -1236,11 +1236,12 @@ class LSU(implicit p: Parameters, edge: TLEdgeOut) extends BoomModule()(p)
                 l_observed) {        // Its only a ordering failure if the cache line was observed between the younger load and us
             ldq_order_fail(i) := true.B
             failed_load := true.B
+            assert(false.B)
```
Comment (jerryz123) [inline src/main/scala/v4/lsu/lsu.scala:1247]: Why is this here?

File: src/main/scala/v4/lsu/lsu.scala [hunk 2]
```diff
@@ -1236,11 +1236,12 @@ class LSU(implicit p: Parameters, edge: TLEdgeOut) extends BoomModule()(p)
...
           }
         } .elsewhen (lcam_ldq_idx(w) =/= i.U) {
           // The load is older, and it wasn't executed
           // we need to kill ourselves, and prevent forwarding
-          when (!(l_executed || l_succeeded)) {
+          when (!(l_executed && l_succeeded)) {
```
Comment (jerryz123) [inline src/main/scala/v4/lsu/lsu.scala]: While I believe this might fix the bug, what is the implication on back-to-back loads that hit in the cache? Will the second load be killed here?

File: src/main/scala/v4/lsu/lsu.scala (hunk 1 above)
Comment (tianrui-wei) [reply src/main/scala/v4/lsu/lsu.scala:1247]: this patch prevents the case that when an older load gets issued, it will not observe a younger load that has been observed. Hence the assertion.

Comment (jerryz123) [reply src/main/scala/v4/lsu/lsu.scala:1247]: Can you add a comment indicating that this case is impossible (and comment out the assignments to ldq_order_fail and failed_load. This will remove the unnecessary logic to set those wires.

File: src/main/scala/v4/lsu/lsu.scala (hunk 2 above)
Comment (tianrui-wei) [reply src/main/scala/v4/lsu/lsu.scala]: That's a good point, I'll think about how to fix it. I think this will prevent back to back loads from working sadly

Comment (tianrui-wei) [reply src/main/scala/v4/lsu/lsu.scala]: I added a fix. Could you take a look at it?

File: src/main/scala/v4/lsu/lsu.scala (hunk 1 above)
Comment (tianrui-wei) [reply src/main/scala/v4/lsu/lsu.scala:1247]: Done.

**Comment by jerryz123** [comment]
Can you make a PR into chipyard updating boom to this version? That will trigger all the Chipyard tests. Also, you can push your branch directly into this repo now

**Comment by tianrui-wei** [comment]
Chipyard PR at https://github.com/ucb-bar/chipyard/pull/2071

File: src/main/scala/v4/lsu/lsu.scala [hunk 3]
```diff
@@ -1270,7 +1281,9 @@ class LSU(implicit p: Parameters, edge: TLEdgeOut) extends BoomModule()(p)
           io.dmem.s1_kill(w) := true.B
         }
         kill_forward(w) := true.B
-      }
+	    }
```
Comment (jerryz123) [inline src/main/scala/v4/lsu/lsu.scala]: Can you fix this brace?

Comment (tianrui-wei) [reply src/main/scala/v4/lsu/lsu.scala]: Fixed!

**Review by jerryz123 — APPROVED** [review]

**Comment by jerryz123** [comment]
FWIW, creating a combinational path from the end of the DCache response stage, into the end of the CAM stage, makes me nervous. It might be ok, but :shrug:

**Comment by tianrui-wei** [comment]
That is a good point. If this ever comes up as a problem, we can delay the kill signal by one cycle and then `and` the signals for kill. I'll merge this in for now and test if this indeed fixes the fastpnr. Do you want me to leave the cy PR open?

File: src/main/scala/v4/lsu/lsu.scala (hunk 2 above)
Comment (Heresyrac) [reply src/main/scala/v4/lsu/lsu.scala]: > While I believe this might fix the bug, what is the implication on back-to-back loads that hit in the cache? Will the second load be killed here?

This implementation has significant  performance issues. I found that testcase where load-store-load operations at the same address are repeated, this modification results in a 26% performance loss. The reason is as follows:

1. Older loads successfully match with previous stores and retrieve data from the STQ. At this point, the executed flag is not set to 1, meaning even though the data has been successfully retrieved, the condition executed && succeed is not met.
2. Younger loads check the LDQ and find an older load at the same address with executed=0, triggering a kill_forward.
The younger load repeatedly goes through wakeup-kill_forward until the older load commits and leaves the LDQ, at which point kill_forward is no longer triggered.
3. The younger load completes, resulting in significant additional latency.
I have already raised a new issue to explain this problem.

Comment (tianrui-wei) [reply src/main/scala/v4/lsu/lsu.scala]: Hi @Heresyrac, thank you for raising this issue. I'll look into this today.
